### PR TITLE
feat: View bundle patches

### PR DESCRIPTION
**Pull request by Ushie**
- [x] Decide whether to use FlowRow or wait for ContextualFlowRow to stabilize
- [x] Whether or not to display patch options
- [x] Minor design adjustments

### CODE REVIEW

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 1]
```diff
@@ -62,44 +61,204 @@ fun BundlePatchesDialog(
...
+                } else {
+                    Row(
+                        horizontalArrangement = Arrangement.spacedBy(4.dp),
+                        verticalAlignment = Alignment.CenterVertically
+                    ) {
+                        PatchInfoChip(
+                            text = "\uD83D\uDCE6 Any package"
+                        )
+                        PatchInfoChip(
+                            text = "\uD83C\uDFAF Any version"
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:183]: This is not translatable

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 2]
```diff
@@ -62,44 +61,204 @@ fun BundlePatchesDialog(
...
+                    val options = patch.options
+
+                    Column {
+                        options.forEachIndexed { i, option ->
+                            OutlinedCard(
+                                modifier = Modifier.fillMaxWidth(), colors = CardColors(
+                                    containerColor = Color.Transparent,
+                                    contentColor = MaterialTheme.colorScheme.onSurface,
+                                    disabledContainerColor = Color.Transparent,
+                                    disabledContentColor = MaterialTheme.colorScheme.onSurface
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:199]: These seem unnecessary if the card cannot be disabled

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 3]
```diff
@@ -62,44 +61,204 @@ fun BundlePatchesDialog(
...
+                )
+            }
+            Column(
+                verticalArrangement = Arrangement.spacedBy(4.dp)
+            ) {
+                if (!patch.compatiblePackages.isNullOrEmpty()) {
+                    patch.compatiblePackages.forEach { compatiblePackage ->
+                        val packageName = compatiblePackage.packageName
+                        val versions = compatiblePackage.versions.orEmpty().reversed()
+                        val itemCount = if (versions.isEmpty()) 1 else versions.size + 1
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:140]: This is unused

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 4]
```diff
@@ -36,13 +35,13 @@ fun BundlePatchesDialog(
     bundle: PatchBundleSource,
 ) {
     var informationCardVisible by remember { mutableStateOf(true) }
+    var showAllVersions by remember { mutableStateOf(false) }
+    var showOptions by remember { mutableStateOf(false) }
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:39]: This state will be lost on configuration changes such as screen rotation. You probably need to use `rememberSaveable` here.

`informationCardVisible` should probably also be `rememberSaveable`.

File: app/src/main/java/app/revanced/manager/ui/component/ArrowButton.kt [hunk 5]
```diff
@@ -13,10 +13,16 @@ import androidx.compose.ui.res.stringResource
 import app.revanced.manager.R
 
 @Composable
-fun ArrowButton(modifier: Modifier = Modifier, expanded: Boolean, onClick: (() -> Unit)?) {
+fun ArrowButton(
+    modifier: Modifier = Modifier,
+    expanded: Boolean,
+    onClick: (() -> Unit)? = null,
+    rotationInitial: Float = 0f,
+    rotationFinal: Float = 180f
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/ArrowButton.kt:21]: Why were these added?

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 6]
```diff
@@ -62,44 +61,204 @@ fun BundlePatchesDialog(
...
+    } else {
+        Modifier
+    }
+
+    OutlinedCard(
+        modifier = modifier.then(cardModifier), colors = CardColors(
+            containerColor = Color.Transparent,
+            contentColor = MaterialTheme.colorScheme.onSurface,
+            disabledContainerColor = Color.Transparent,
+            disabledContentColor = MaterialTheme.colorScheme.onSurface
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:247]: Same here

**Review by Axelen123 — CHANGES_REQUESTED** [review]

File: gradle/libs.versions.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [versions]
 ktx = "1.13.1"
-material3 = "1.2.1"
+material3 = "1.3.0-beta05"
```
[the remark this answers is not in the document]
Comment (Ushie) [reply gradle/libs.versions.toml:3]: Not anymore but upon downgrading back to 1.2.1 for me the app kept crashing (most likely local issue) so I just kept it

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt (hunk 2 above)
Comment (Ushie) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:199]: These parameters are not optional/nullable

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt (hunk 6 above)
Comment (Ushie) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:247]: These parameters are not optional/nullable

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 8]
```diff
@@ -62,44 +63,204 @@ fun BundlePatchesDialog(
...
             ) {
-                item {
-                    AnimatedVisibility(visible = informationCardVisible) {
-                        NotificationCard(
-                            icon = Icons.Outlined.Lightbulb,
-                            text = stringResource(R.string.tap_on_patches),
-                            onDismiss = { informationCardVisible = false }
+                state.patchBundleOrNull()?.let { bundle ->
+                    items(bundle.patches.size) { bundleIndex ->
+                        val patch = bundle.patches[bundleIndex]
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:72]: There is an overload for `List`s. I don't know why this wasn't being used in the original file. You may need to import it.

```suggestion
                    items(bundle.patches) { patch ->
```

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 9]
```diff
@@ -1,48 +1,49 @@
...
+    var showOptions by rememberSaveable { mutableStateOf(false) }
     val state by bundle.state.collectAsStateWithLifecycle()
 
     Dialog(
-        onDismissRequest = onDismissRequest,
-        properties = DialogProperties(
-            usePlatformDefaultWidth = false,
-            dismissOnBackPress = true
+        onDismissRequest = onDismissRequest, properties = DialogProperties(
+            usePlatformDefaultWidth = false, dismissOnBackPress = true
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:46]:
```suggestion
        onDismissRequest = onDismissRequest,
        properties = DialogProperties(
            usePlatformDefaultWidth = false,
            dismissOnBackPress = true
```

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 10]
```diff
@@ -62,44 +63,204 @@ fun BundlePatchesDialog(
...
-                item {
-                    AnimatedVisibility(visible = informationCardVisible) {
-                        NotificationCard(
-                            icon = Icons.Outlined.Lightbulb,
-                            text = stringResource(R.string.tap_on_patches),
-                            onDismiss = { informationCardVisible = false }
+                state.patchBundleOrNull()?.let { bundle ->
+                    items(bundle.patches.size) { bundleIndex ->
+                        val patch = bundle.patches[bundleIndex]
+                        PatchItem(patch,
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:73]: Add a newline

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 11]
```diff
@@ -62,44 +63,204 @@ fun BundlePatchesDialog(
...
+        modifier = Modifier
+            .fillMaxWidth()
+            .then(
+                if (patch.options.isNullOrEmpty()) Modifier else Modifier
+                    .clip(RoundedCornerShape(8.dp))
+                    .clickable(onClick = onExpandOptions),
+            )
+    ) {
+        Column(
+            modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(6.dp)
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:105]:
```suggestion
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(6.dp)
```

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 12]
```diff
@@ -62,44 +61,204 @@ fun BundlePatchesDialog(
...
+                    Column {
+                        options.forEachIndexed { i, option ->
+                            OutlinedCard(
+                                modifier = Modifier.fillMaxWidth(), colors = CardColors(
+                                    containerColor = Color.Transparent,
+                                    contentColor = MaterialTheme.colorScheme.onSurface,
+                                    disabledContainerColor = Color.Transparent,
+                                    disabledContentColor = MaterialTheme.colorScheme.onSurface
+                                ), shape = when {
+                                    i == 0 && options.lastIndex == 0 -> RoundedCornerShape(8.dp)
```
[the remark this answers is not in the document]
Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:195]: This is still here

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 13]
```diff
@@ -62,44 +63,204 @@ fun BundlePatchesDialog(
...
+                    fontWeight = FontWeight.Bold
+                )
+
+                if (!patch.options.isNullOrEmpty()) {
+                    ArrowButton(expanded = expandOptions)
+                }
+            }
+            patch.description?.let {
+                Text(
+                    text = it, style = MaterialTheme.typography.bodyMedium
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:125]:
```suggestion
                    text = it,
                    style = MaterialTheme.typography.bodyMedium
```

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 14]
```diff
@@ -62,44 +63,204 @@ fun BundlePatchesDialog(
...
+                                    modifier = Modifier.padding(12.dp),
+                                    verticalArrangement = Arrangement.spacedBy(4.dp),
+                                ) {
+                                    Text(
+                                        text = option.title,
+                                        style = MaterialTheme.typography.titleMedium,
+                                        color = MaterialTheme.colorScheme.primary
+                                    )
+                                    Text(
+                                        text = option.description, style = MaterialTheme.typography.bodyMedium
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:211]:
```suggestion
                                        text = option.description,
                                        style = MaterialTheme.typography.bodyMedium
```

File: gradle/libs.versions.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (Axelen123) [reply gradle/libs.versions.toml:3]: What does the crash look like?

File: app/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -372,4 +372,7 @@
     <string name="never_show_again">Never show again</string>
     <string name="show_manager_update_dialog_on_launch">Show update message on launch</string>
     <string name="show_manager_update_dialog_on_launch_description">Shows a popup notification whenever there is a new update available on launch.</string>
+    <string name="less">Less</string>
```
Comment (Axelen123) [inline app/src/main/res/values/strings.xml:375]: Move this below the string with the name "more".

File: app/src/main/res/values/strings.xml [hunk 16]
```diff
@@ -372,4 +372,7 @@
...
+    <string name="any_version">Any version</string>
+    <string name="any_package">Any package</string>
```
Comment (Axelen123) [inline app/src/main/res/values/strings.xml:377]: Move and rename these to make things consistent with the other bundle dialog strings.

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 17]
```diff
@@ -62,44 +63,204 @@ fun BundlePatchesDialog(
...
+    val cardModifier = if (onClick != null) {
+        Modifier
+            .clip(shape)
+            .clickable(onClick = onClick)
+    } else {
+        Modifier
+    }
+
+    OutlinedCard(
+        modifier = modifier.then(cardModifier), colors = CardColors(
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:240]:
```suggestion
        modifier = modifier.then(cardModifier),
        colors = CardColors(
```

File: app/src/main/java/app/revanced/manager/ui/component/ArrowButton.kt (hunk 5 above)
Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/ui/component/ArrowButton.kt:21]: These changes are still here. I assume this is unintentional.

**Review by Axelen123 — CHANGES_REQUESTED** [review]

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt (hunk 8 above)
Comment (Ushie) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:72]: There's no such thing, but this is the next best thing:

```kotlin
items(
    count = bundle.patches.size,
    key = { index -> bundle.patches[index] }
) { bundleIndex ->
    val patch = bundle.patches[bundleIndex]
    
    PatchItem(
        patch,
        showAllVersions,
        onExpandVersions = { showAllVersions = !showAllVersions },
        showOptions,
        onExpandOptions = { showOptions = !showOptions }
    )
}

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ushie) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:195]: Yep, Jetbrain's Apply Locally seems to be broken

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt (hunk 8 above)
Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt]: There is: https://developer.android.com/reference/kotlin/androidx/compose/foundation/lazy/LazyListScope#(androidx.compose.foundation.lazy.LazyListScope).items(kotlin.collections.List,kotlin.Function1,kotlin.Function1,kotlin.Function2)

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 18]
```diff
@@ -1,48 +1,49 @@
...
+    var showAllVersions by rememberSaveable { mutableStateOf(false) }
+    var showOptions by rememberSaveable { mutableStateOf(false) }
     val state by bundle.state.collectAsStateWithLifecycle()
 
     Dialog(
         onDismissRequest = onDismissRequest,
         properties = DialogProperties(
-            usePlatformDefaultWidth = false,
-            dismissOnBackPress = true
+            usePlatformDefaultWidth = false, dismissOnBackPress = true
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:46]: Revert this.

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 19]
```diff
@@ -62,44 +63,213 @@ fun BundlePatchesDialog(
...
+            patch.description?.let {
+                Text(
+                    text = it,
+                    style = MaterialTheme.typography.bodyMedium
+                )
+            }
+            Column(
+                verticalArrangement = Arrangement.spacedBy(4.dp)
+            ) {
+                if (!patch.compatiblePackages.isNullOrEmpty()) {
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:138]: Reverse the condition and swap the `if` and `else` blocks

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt [hunk 20]
```diff
@@ -62,44 +63,213 @@ fun BundlePatchesDialog(
...
+                    patch.compatiblePackages.forEach { compatiblePackage ->
+                        val packageName = compatiblePackage.packageName
+                        val versions = compatiblePackage.versions.orEmpty().reversed()
```
Comment (Axelen123) [inline app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:141]: Why are the versions reversed?

**Review by Axelen123 — CHANGES_REQUESTED** [review]

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt (hunk 20 above)
Comment (Ushie) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:141]: Because when patch devs add new versions, they add them to the end of the list rather than the beginning

File: app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt (hunk 18 above)
Comment (Ushie) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:46]: God I've undone that a thousand times by now, I have removing new lines disabled in Intellij I don't understand why that's happening

Comment (Axelen123) [reply app/src/main/java/app/revanced/manager/ui/component/bundle/BundlePatchesDialog.kt:46]: Maybe try another editor for formatting fixes? (just make sure it doesn't switch the file to CRLF or something)

**Review by Axelen123 — COMMENTED** [review]
The "patch options are expanded" state is shared for all cards. Is this intentional or not?

**Comment by Ushie** [comment]
Yes
